ext/drx/drx.c

```diff
@@ -129,13 +137,20 @@ bool
 drx_init(void)
 {
     /* drx_insert_counter_update() needs 1 slot on x86 plus the 1 slot
-       drreg uses for aflags, and 2 reg slots on aarch, so 2 on both.
+     * drreg uses for aflags, and 2 reg slots on aarch, so 2 on both.
+     * drx_expand_scatter_gather() needs 4 slots in app2app phase, which
+     * cannot be reused by other phases. So, ideally we should reserve
+     * 6 slots for drx. But we settle with 4 to avoid stealing too many
+     * slots from other clients/libs. When drx needs more for instrumenting
+     * scatter/gather instrs, we fall back on DR slots. As scatter/gather
```

Comment:
I forgot about this limitation: DR spill slots are not allowed to be used across application instructions.  I think drreg has logic to not do lazy restores for DR spill slots so there is no correctness issue.  (Though in my head I'm worried about clients who have custom slots they use across app instrs...but those should be disjoint from the slots drreg has reserved...)  But the cost is not just the indirection but the extra restores, which can be significant.  Hmm.  Given that this is limited to scatter-gather single-instr bb's: maybe it is fine and all my intuition-directed misgivings are about other TLS slots and not drreg slots.

Reply:
I wonder if we could do some performance measurements to reassure ourselves.

Reply:
>  DR spill slots are not allowed to be used across application instructions. I think drreg has logic to not do lazy restores for DR spill slots so there is no correctness issue. 

Another reason for there not being any cross-app instr usage of DR slots here:
drreg restores all slot vals at bb end. As scatter/gather instrs have their own separate bb (with only a single instr), the DR slot would not be used beyond the scatter/gather app instr anyway. 

> But the cost is not just the indirection but the extra restores, which can be significant.
> I wonder if we could do some performance measurements to reassure ourselves.

Checking my understanding here: there are multiple components to the additional cost, most of which apply only for scatter/gather instrs. Which is the one we are concerned about?:
- indirection in use of DR slots: if no scatter/gather instr, no DR slots would need to be used as drx reserves atleast as many TLS slots as before in this PR.
- drx using more TLS slots for scatter/gather expansion, making them less available to other clients: Again, if no scatter/gather instr, drx won't use any more TLS slots than before. So, other clients will see same # of TLS slots via drreg.
- reservation of more TLS slots via `dr_raw_tls_calloc`, making them less available to the application: This cost is present even if there aren't any scatter/gather instrs in the app. 
  - For Windows, there'd be no additional cost as we're requesting the same number of TLS slots as before.
  - For Linux, where we do request more TLS slots now: does this cost apply to Linux too? [1] isn't clear about whether TLS slots are shared with apps on Linux too. It only mentions that they are plentiful on Linux and only have a client-total limit of 64 -- are you concerned about hitting that limit?

About performance measurements: I'm trying to think what the baseline would be. If my above understanding is correct, all additional cost is when we see scatter/gather instrs, which weren't supported before now.



[1]: https://github.com/DynamoRIO/dynamorio/blob/10aed67a4e2d0ecd7272101543873b4ecfab78f9/core/lib/dr_tools.h#L1975

Reply:
> As scatter/gather instrs have their own separate bb (with only a single instr), the DR slot would not be used beyond the scatter/gather app instr anyway.

Aren't the expanded loads/stores marked as app instrs?  I would assume the bb ends up with a bunch of app instrs, let's say 8.  drx's app2app now takes the direct-TLS slots, and the client sticks in some complex instrumentation in the insertion phase which now has to restore and re-spill around each of those 8 app instrs: so if it's using 3 slots that's 24 extra loads and stores.  OTOH the scatter expansion is pretty heavyweight already.

Mostly I recall going to great lengths to tune things to not use DR slots, and problems when that tuning failed.  Dr. Memory is the biggest most complex client we have, so I would say let's test on it: but its switch to use drreg is not finished (and may never be...).  I just "have a bad feeling".  But not enough concrete details to hold this up.

Reply:
> For Windows, there'd be no additional cost as we're requesting the same number of TLS slots as before.

No, we're asking for 2 additional for clients who are already using 2 (which most are), b/c of the change to do_not_sum_slots.  Perhaps that should remain true for Windows.

Reply:
> Aren't the expanded loads/stores marked as app instrs? 

Oh you're right. So drreg's handling to skip lazy restore for DR slots saves the day then.

> No, we're asking for 2 additional for clients who are already using 2 (which most are), b/c of the change to do_not_sum_slots. Perhaps that should remain true for Windows.

Done

Reply:
Discussed offline that we'll need to observe perf impact and fine-tune slot reservations if needed. There are unknowns in its impact on Windows and large clients like DrMemory. Good to submit for now though.